From 2747f70ed652fd69e6aaeab9a1546d2636cca0ac Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Tue, 18 Jun 2013 18:58:16 +0300
Subject: i.MX6: gpio: fix gpio read mode
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch fixes gpio read mode.
This patch takes into account the GPIO direction in order to calculate a correct value for reading the GPIO value.
Without this path the “mxc_gpio_get” function returned correct values for input GPIOs only.
---
 arch/arm/plat-mxc/gpio.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/plat-mxc/gpio.c b/arch/arm/plat-mxc/gpio.c
index 82b70d5..242b02b 100755
--- a/arch/arm/plat-mxc/gpio.c
+++ b/arch/arm/plat-mxc/gpio.c
@@ -284,7 +284,10 @@ static int mxc_gpio_get(struct gpio_chip *chip, unsigned offset)
 	struct mxc_gpio_port *port =
 		container_of(chip, struct mxc_gpio_port, chip);
 
-	return (__raw_readl(port->base + GPIO_PSR) >> offset) & 1;
+	unsigned int base_offset = ((__raw_readl(port->base + GPIO_GDIR) & 
+								 (1 << offset)) ? GPIO_DR : GPIO_PSR);
+
+	return (__raw_readl(port->base + base_offset) >> offset) & 1;
 }
 
 static int mxc_gpio_direction_input(struct gpio_chip *chip, unsigned offset)
-- 
1.7.9.5

